File: infra/generate_icons.mjs
/**
 * generate_icons.mjs — LifePlan app icon generator (pure Node.js, no dependencies)
 *
 * Generates Android launcher PNGs at all required densities.
 * Run from project root:
 *   node infra/generate_icons.mjs
 */

import { createWriteStream, mkdirSync } from "fs";
import { deflateSync } from "zlib";
import { dirname, join } from "path";
import { fileURLToPath } from "url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const PROJECT_ROOT = join(__dirname, "..");
const RES = join(PROJECT_ROOT, "frontend/android/app/src/main/res");

// ─── Output targets ────────────────────────────────────────────────────────
const LAUNCHER_TARGETS = [
  [48,  "mipmap-mdpi"],
  [72,  "mipmap-hdpi"],
  [96,  "mipmap-xhdpi"],
  [144, "mipmap-xxhdpi"],
  [192, "mipmap-xxxhdpi"],
];

// ─── Design constants (at 1024px scale) ───────────────────────────────────
const CANVAS = 1024;
const BG_STOPS = [
  [0.00, [55,  48,  163]],   // #3730a3 deep indigo
  [0.65, [79,  62,  184]],   // mid blend
  [0.88, [124, 58,  237]],   // #7c3aed violet-purple
  [1.00, [91,  33,  182]],   // #5b21b6 darker ring
];
const ARM_HW   = 40;   // arm half-width
const ARM_VR   = 400;  // vertical arm reach from center
const ARM_HR   = 370;  // horizontal arm reach from center
const ARM_IO   = 60;   // inner offset (arm starts here)
const CR       = 55;   // center circle radius
const TDR      = 20;   // tip dot radius
const MARK_A   = 0.92; // mark opacity

// ─── PNG encoder ─────────────────────────────────────────────────────────
function crc32(buf) {
  const table = crc32.table || (crc32.table = (() => {
    const t = new Uint32Array(256);
    for (let i = 0; i < 256; i++) {
      let c = i;
      for (let k = 0; k < 8; k++) c = (c & 1) ? (0xEDB88320 ^ (c >>> 1)) : (c >>> 1);
      t[i] = c;
    }
    return t;
  })());
  let c = 0xFFFFFFFF;
  for (const b of buf) c = table[(c ^ b) & 0xFF] ^ (c >>> 8);
  return (c ^ 0xFFFFFFFF) >>> 0;
}

function chunk(type, data) {
  const typeBuf = Buffer.from(type, "ascii");
  const len = Buffer.alloc(4); len.writeUInt32BE(data.length);
  const crcBuf = Buffer.alloc(4);
  crcBuf.writeUInt32BE(crc32(Buffer.concat([typeBuf, data])));
  return Buffer.concat([len, typeBuf, data, crcBuf]);
}

function encodePNG(pixels, width, height) {
  // pixels: Uint8ClampedArray of RGBA, row-major
  const sig = Buffer.from([137, 80, 78, 71, 13, 10, 26, 10]);

  const ihdr = Buffer.alloc(13);
  ihdr.writeUInt32BE(width,  0);
  ihdr.writeUInt32BE(height, 4);
  ihdr[8]  = 8;  // bit depth
  ihdr[9]  = 6;  // RGBA color type
  ihdr[10] = 0; ihdr[11] = 0; ihdr[12] = 0;

  // Build filtered scanlines (filter byte 0 = None per row)
  const rowLen = width * 4;
  const raw = Buffer.alloc(height * (rowLen + 1));
  for (let y = 0; y < height; y++) {
    raw[y * (rowLen + 1)] = 0; // filter type None
    for (let x = 0; x < rowLen; x++) {
      raw[y * (rowLen + 1) + 1 + x] = pixels[y * rowLen + x];
    }
  }

  const compressed = deflateSync(raw, { level: 6 });
  return Buffer.concat([
    sig,
    chunk("IHDR", ihdr),
    chunk("IDAT", compressed),
    chunk("IEND", Buffer.alloc(0)),
  ]);
}

// ─── Drawing helpers ──────────────────────────────────────────────────────
function lerp(a, b, t) { return a + (b - a) * t; }

function lerpColor(c1, c2, t) {
  return [lerp(c1[0], c2[0], t), lerp(c1[1], c2[1], t), lerp(c1[2], c2[2], t)];
}

function gradientAt(t, stops) {
  t = Math.max(0, Math.min(1, t));
  for (let i = 0; i < stops.length - 1; i++) {
    const [t0, c0] = stops[i];
    const [t1, c1] = stops[i + 1];
    if (t >= t0 && t <= t1) {
      const lt = t1 === t0 ? 0 : (t - t0) / (t1 - t0);
      return lerpColor(c0, c1, lt);
    }
  }
  return stops[stops.length - 1][1];
}

// ─── Render full icon ─────────────────────────────────────────────────────
function renderIcon(size) {
  const sc = size / CANVAS;
  const pixels = new Uint8ClampedArray(size * size * 4);
  const cx = size / 2;
  const cy = size / 2;

  // Helper: set RGBA at (x,y) with alpha-over blend
  function blendPixel(px, py, r, g, b, a) {
    if (px < 0 || py < 0 || px >= size || py >= size) return;
    const idx = (Math.round(py) * size + Math.round(px)) * 4;
    const af = a / 255;
    const ar = pixels[idx + 3] / 255;
    const ao = af + ar * (1 - af);
    if (ao < 1e-6) return;
    pixels[idx]     = Math.round((r * af + pixels[idx]   * ar * (1 - af)) / ao);
    pixels[idx + 1] = Math.round((g * af + pixels[idx+1] * ar * (1 - af)) / ao);
    pixels[idx + 2] = Math.round((b * af + pixels[idx+2] * ar * (1 - af)) / ao);
    pixels[idx + 3] = Math.round(ao * 255);
  }

  // ── Background radial gradient ──
  const maxR = size / 2;
  for (let y = 0; y < size; y++) {
    for (let x = 0; x < size; x++) {
      const dx = x - cx, dy = y - cy;
      const t = Math.sqrt(dx * dx + dy * dy) / maxR;
      const [r, g, b] = gradientAt(t, BG_STOPS);
      const idx = (y * size + x) * 4;
      pixels[idx]   = Math.round(r);
      pixels[idx+1] = Math.round(g);
      pixels[idx+2] = Math.round(b);
      pixels[idx+3] = 255;
    }
  }

  // ── Glow (soft white halo at center) ──
  const glowR = 160 * sc;
  const glowA = 0.18;
  for (let y = 0; y < size; y++) {
    for (let x = 0; x < size; x++) {
      const dx = x - cx, dy = y - cy;
      const dist = Math.sqrt(dx * dx + dy * dy);
      if (dist < glowR) {
        const t = 1 - dist / glowR;
        const a = Math.round(glowA * t * t * 255);
        blendPixel(x, y, 255, 255, 255, a);
      }
    }
  }

  // ── Mark drawing ──
  const markA = Math.round(MARK_A * 255);
  function s(v) { return v * sc; }

  // Draw filled anti-aliased rounded rectangle via pixel scan
  function fillRoundRect(x1, y1, x2, y2, radius) {
    const r = Math.min(radius, Math.abs(x2 - x1) / 2, Math.abs(y2 - y1) / 2);
    const X1 = Math.min(x1, x2), Y1 = Math.min(y1, y2);
    const X2 = Math.max(x1, x2), Y2 = Math.max(y1, y2);
    for (let py = Math.floor(Y1); py <= Math.ceil(Y2); py++) {
      for (let px = Math.floor(X1); px <= Math.ceil(X2); px++) {
        // Check if pixel center is inside rounded rect
        const inside = isInsideRoundRect(px + 0.5, py + 0.5, X1, Y1, X2, Y2, r);
        if (inside > 0) {
          blendPixel(px, py, 255, 255, 255, Math.round(markA * inside));
        }
      }
    }
  }

  function isInsideRoundRect(px, py, x1, y1, x2, y2, r) {
    // Returns coverage [0,1]
    if (px < x1 || px > x2 || py < y1 || py > y2) return 0;
    // Corner checks
    const corners = [
      [x1 + r, y1 + r], [x2 - r, y1 + r],
      [x1 + r, y2 - r], [x2 - r, y2 - r],
    ];
    for (const [cx2, cy2] of corners) {
      if (px < x1 + r && py < y1 + r && (px === cx2 && py === cy2 ? 0 : 1)) {
        // top-left corner
        if (px < x1 + r && py < y1 + r) {
          const d = Math.sqrt((px - (x1+r))**2 + (py - (y1+r))**2);
          if (d > r + 1) return 0;
          if (d > r - 1) return Math.max(0, r + 0.5 - d);
          break;
        }
      }
    }
    // Simplified: use distance to nearest corner circle
    const nearX = Math.max(x1 + r, Math.min(x2 - r, px));
    const nearY = Math.max(y1 + r, Math.min(y2 - r, py));
    const dx = px - nearX, dy = py - nearY;
    if (dx === 0 && dy === 0) return 1; // fully inside
    const d = Math.sqrt(dx * dx + dy * dy);
    if (d > r + 0.7) return 0;
    if (d < r - 0.7) return 1;
    return Math.max(0, Math.min(1, (r + 0.5 - d)));
  }

  function fillCircle(ocx, ocy, radius) {
    for (let py = Math.floor(ocy - radius - 1); py <= Math.ceil(ocy + radius + 1); py++) {
      for (let px = Math.floor(ocx - radius - 1); px <= Math.ceil(ocx + radius + 1); px++) {
        const d = Math.sqrt((px + 0.5 - ocx) ** 2 + (py + 0.5 - ocy) ** 2);
        if (d < radius + 0.7) {
          const coverage = Math.max(0, Math.min(1, radius + 0.5 - d));
          blendPixel(px, py, 255, 255, 255, Math.round(markA * coverage));
        }
      }
    }
  }

  // Top arm
  fillRoundRect(cx - s(ARM_HW), cy - s(ARM_VR), cx + s(ARM_HW), cy - s(ARM_IO), s(ARM_HW));
  // Bottom arm
  fillRoundRect(cx - s(ARM_HW), cy + s(ARM_IO), cx + s(ARM_HW), cy + s(ARM_VR), s(ARM_HW));
  // Left arm
  fillRoundRect(cx - s(ARM_HR), cy - s(ARM_HW), cx - s(ARM_IO), cy + s(ARM_HW), s(ARM_HW));
  // Right arm
  fillRoundRect(cx + s(ARM_IO), cy - s(ARM_HW), cx + s(ARM_HR), cy + s(ARM_HW), s(ARM_HW));
  // Center circle
  fillCircle(cx, cy, s(CR));
  // Tip dots
  fillCircle(cx, cy - s(ARM_VR - ARM_HW), s(TDR));
  fillCircle(cx, cy + s(ARM_VR - ARM_HW), s(TDR));
  fillCircle(cx - s(ARM_HR - ARM_HW), cy, s(TDR));
  fillCircle(cx + s(ARM_HR - ARM_HW), cy, s(TDR));

  return pixels;
}

// ─── Also render foreground-only (transparent bg) ─────────────────────────
function renderForeground(size) {
  const sc = size / CANVAS;
  const pixels = new Uint8ClampedArray(size * size * 4); // starts transparent
  const cx = size / 2;
  const cy = size / 2;
  const markA = Math.round(MARK_A * 255);
  function s(v) { return v * sc; }

  function blendPixel(px, py, r, g, b, a) {
    if (px < 0 || py < 0 || px >= size || py >= size) return;
    const idx = (Math.round(py) * size + Math.round(px)) * 4;
    const af = a / 255;
    const ar = pixels[idx + 3] / 255;
    const ao = af + ar * (1 - af);
    if (ao < 1e-6) return;
    pixels[idx]     = Math.round((r * af + pixels[idx]   * ar * (1 - af)) / ao);
    pixels[idx + 1] = Math.round((g * af + pixels[idx+1] * ar * (1 - af)) / ao);
    pixels[idx + 2] = Math.round((b * af + pixels[idx+2] * ar * (1 - af)) / ao);
    pixels[idx + 3] = Math.round(ao * 255);
  }

  function isInsideRoundRect(px, py, x1, y1, x2, y2, r) {
    if (px < x1 || px > x2 || py < y1 || py > y2) return 0;
    const nearX = Math.max(x1 + r, Math.min(x2 - r, px));
    const nearY = Math.max(y1 + r, Math.min(y2 - r, py));
    const dx = px - nearX, dy = py - nearY;
    if (dx === 0 && dy === 0) return 1;
    const d = Math.sqrt(dx * dx + dy * dy);
    if (d > r + 0.7) return 0;
    if (d < r - 0.7) return 1;
    return Math.max(0, Math.min(1, (r + 0.5 - d)));
  }

  function fillRoundRect(x1, y1, x2, y2, radius) {
    const r = Math.min(radius, Math.abs(x2 - x1) / 2, Math.abs(y2 - y1) / 2);
    const X1 = Math.min(x1, x2), Y1 = Math.min(y1, y2);
    const X2 = Math.max(x1, x2), Y2 = Math.max(y1, y2);
    for (let py = Math.floor(Y1); py <= Math.ceil(Y2); py++) {
      for (let px = Math.floor(X1); px <= Math.ceil(X2); px++) {
        const inside = isInsideRoundRect(px + 0.5, py + 0.5, X1, Y1, X2, Y2, r);
        if (inside > 0) blendPixel(px, py, 255, 255, 255, Math.round(markA * inside));
      }
    }
  }

  function fillCircle(ocx, ocy, radius) {
    for (let py = Math.floor(ocy - radius - 1); py <= Math.ceil(ocy + radius + 1); py++) {
      for (let px = Math.floor(ocx - radius - 1); px <= Math.ceil(ocx + radius + 1); px++) {
        const d = Math.sqrt((px + 0.5 - ocx) ** 2 + (py + 0.5 - ocy) ** 2);
        if (d < radius + 0.7) {
          const coverage = Math.max(0, Math.min(1, radius + 0.5 - d));
          blendPixel(px, py, 255, 255, 255, Math.round(markA * coverage));
        }
      }
    }
  }

  fillRoundRect(cx - s(ARM_HW), cy - s(ARM_VR), cx + s(ARM_HW), cy - s(ARM_IO), s(ARM_HW));
  fillRoundRect(cx - s(ARM_HW), cy + s(ARM_IO), cx + s(ARM_HW), cy + s(ARM_VR), s(ARM_HW));
  fillRoundRect(cx - s(ARM_HR), cy - s(ARM_HW), cx - s(ARM_IO), cy + s(ARM_HW), s(ARM_HW));
  fillRoundRect(cx + s(ARM_IO), cy - s(ARM_HW), cx + s(ARM_HR), cy + s(ARM_HW), s(ARM_HW));
  fillCircle(cx, cy, s(CR));
  fillCircle(cx, cy - s(ARM_VR - ARM_HW), s(TDR));
  fillCircle(cx, cy + s(ARM_VR - ARM_HW), s(TDR));
  fillCircle(cx - s(ARM_HR - ARM_HW), cy, s(TDR));
  fillCircle(cx + s(ARM_HR - ARM_HW), cy, s(TDR));

  return pixels;
}

// ─── Save PNG ─────────────────────────────────────────────────────────────
function savePNG(pixels, width, height, filePath) {
  mkdirSync(dirname(filePath), { recursive: true });
  const pngBuf = encodePNG(pixels, width, height);
  const ws = createWriteStream(filePath);
  ws.write(pngBuf);
  ws.end();
  const rel = filePath.replace(PROJECT_ROOT + "/", "").replace(PROJECT_ROOT + "\\", "");
  console.log(`  Saved: ${rel}  (${width}x${height})`);
}

// ─── Main ─────────────────────────────────────────────────────────────────
console.log("LifePlan Icon Generator");
console.log("=".repeat(50));

console.log("\nGenerating launcher icons...");
for (const [size, folder] of LAUNCHER_TARGETS) {
  console.log(`  Rendering ${size}x${size}...`);
  const pixels = renderIcon(size);
  const dir = join(RES, folder);
  savePNG(pixels, size, size, join(dir, "ic_launcher.png"));
  savePNG(pixels, size, size, join(dir, "ic_launcher_round.png"));
}

console.log("\nGenerating adaptive icon foreground (1024x1024, transparent bg)...");
const fgPixels = renderForeground(256); // 256 is enough for foreground layer
savePNG(fgPixels, 256, 256, join(RES, "drawable", "ic_launcher_foreground.png"));

console.log("\nDone.");

File: infra/generate_icons.py
"""
generate_icons.py — Loadout app icon generator

Generates Android launcher PNGs at all required densities using Pillow.
No SVG renderer required — the icon is drawn directly via PIL drawing primitives.

Run from the project root:
    python infra/generate_icons.py

Requires:
    pip install Pillow
"""

import os
import math
from pathlib import Path
from PIL import Image, ImageDraw, ImageFilter

# ---------------------------------------------------------------------------
# Project root — resolve relative to this script's location
# ---------------------------------------------------------------------------
SCRIPT_DIR = Path(__file__).resolve().parent
PROJECT_ROOT = SCRIPT_DIR.parent
RES_DIR = PROJECT_ROOT / "frontend" / "android" / "app" / "src" / "main" / "res"

# ---------------------------------------------------------------------------
# Output targets
# ---------------------------------------------------------------------------
LAUNCHER_TARGETS = [
    (48,   RES_DIR / "mipmap-mdpi"),
    (72,   RES_DIR / "mipmap-hdpi"),
    (96,   RES_DIR / "mipmap-xhdpi"),
    (144,  RES_DIR / "mipmap-xxhdpi"),
    (192,  RES_DIR / "mipmap-xxxhdpi"),
]

FOREGROUND_TARGET = (1024, RES_DIR / "drawable")

# ---------------------------------------------------------------------------
# Design constants (all defined at 1024px scale, scaled linearly)
# ---------------------------------------------------------------------------
CANVAS = 1024

# Background gradient stops  (indigo center → violet edges → subtle darker ring)
BG_GRADIENT_STOPS = [
    (0.00, (55,  48, 163)),   # #3730a3  deep indigo
    (0.65, (79,  62, 184)),   # midpoint blend
    (0.88, (124, 58, 237)),   # #7c3aed  violet-purple
    (1.00, (91,  33, 182)),   # #5b21b6  darker outer ring
]

# Mark dimensions (at 1024px scale)
ARM_HALF_WIDTH   = 40    # half of 80px arm width
ARM_V_REACH      = 400   # top/bottom arm: tip distance from center (vertical)
ARM_H_REACH      = 370   # left/right arm: tip distance from center (horizontal)
ARM_INNER_OFFSET = 60    # arm starts this far from center
CENTER_RADIUS    = 55    # center circle radius
TIP_DOT_RADIUS   = 20    # tip dot radius
MARK_OPACITY     = 0.90  # white mark opacity

# Glow parameters (at 1024px scale)
GLOW_RADIUS      = 180   # radial extent of inner glow
GLOW_BLUR        = 60    # gaussian blur sigma for glow
GLOW_OPACITY     = 0.22  # peak opacity of glow


# ---------------------------------------------------------------------------
# Helper: lerp between two RGB colours
# ---------------------------------------------------------------------------
def lerp_color(c1, c2, t):
    return tuple(int(c1[i] + (c2[i] - c1[i]) * t) for i in range(3))


def gradient_color_at(t, stops):
    """Return interpolated RGB for position t in [0,1] given gradient stops."""
    t = max(0.0, min(1.0, t))
    for i in range(len(stops) - 1):
        t0, c0 = stops[i]
        t1, c1 = stops[i + 1]
        if t0 <= t <= t1:
            local_t = (t - t0) / (t1 - t0) if t1 != t0 else 0
            return lerp_color(c0, c1, local_t)
    return stops[-1][1]


# ---------------------------------------------------------------------------
# Build radial gradient background image (RGBA, size x size)
# ---------------------------------------------------------------------------
def make_background(size):
    img = Image.new("RGBA", (size, size), (0, 0, 0, 0))
    pixels = img.load()
    cx = cy = size / 2.0
    max_r = size / 2.0
    for y in range(size):
        for x in range(size):
            dx = x - cx
            dy = y - cy
            r = math.sqrt(dx * dx + dy * dy)
            t = r / max_r
            color = gradient_color_at(t, BG_GRADIENT_STOPS)
            pixels[x, y] = color + (255,)
    return img


# ---------------------------------------------------------------------------
# Build the compass mark as an RGBA image (white mark on transparent bg)
# size: output pixel dimension
# Returns RGBA image with the compass mark only.
# ---------------------------------------------------------------------------
def make_mark(size):
    scale = size / CANVAS
    img = Image.new("RGBA", (size, size), (0, 0, 0, 0))
    draw = ImageDraw.Draw(img)

    cx = cy = size / 2.0
    alpha = int(255 * MARK_OPACITY)
    white = (255, 255, 255, alpha)

    def s(v):
        """Scale a design value from 1024-canvas to output size."""
        return v * scale

    # ── Vertical arm (top) ──────────────────────────────────────────────────
    # Spans from (cy - ARM_V_REACH) to (cy - ARM_INNER_OFFSET), centred on cx
    top_arm_rect = [
        cx - s(ARM_HALF_WIDTH),
        cy - s(ARM_V_REACH),
        cx + s(ARM_HALF_WIDTH),
        cy - s(ARM_INNER_OFFSET),
    ]
    draw.rounded_rectangle(top_arm_rect, radius=s(ARM_HALF_WIDTH), fill=white)

    # ── Vertical arm (bottom) ───────────────────────────────────────────────
    bot_arm_rect = [
        cx - s(ARM_HALF_WIDTH),
        cy + s(ARM_INNER_OFFSET),
        cx + s(ARM_HALF_WIDTH),
        cy + s(ARM_V_REACH),
    ]
    draw.rounded_rectangle(bot_arm_rect, radius=s(ARM_HALF_WIDTH), fill=white)

    # ── Horizontal arm (left) ───────────────────────────────────────────────
    left_arm_rect = [
        cx - s(ARM_H_REACH),
        cy - s(ARM_HALF_WIDTH),
        cx - s(ARM_INNER_OFFSET),
        cy + s(ARM_HALF_WIDTH),
    ]
    draw.rounded_rectangle(left_arm_rect, radius=s(ARM_HALF_WIDTH), fill=white)

    # ── Horizontal arm (right) ──────────────────────────────────────────────
    right_arm_rect = [
        cx + s(ARM_INNER_OFFSET),
        cy - s(ARM_HALF_WIDTH),
        cx + s(ARM_H_REACH),
        cy + s(ARM_HALF_WIDTH),
    ]
    draw.rounded_rectangle(right_arm_rect, radius=s(ARM_HALF_WIDTH), fill=white)

    # ── Center circle ───────────────────────────────────────────────────────
    cr = s(CENTER_RADIUS)
    draw.ellipse([cx - cr, cy - cr, cx + cr, cy + cr], fill=white)

    # ── Tip dots ────────────────────────────────────────────────────────────
    tr = s(TIP_DOT_RADIUS)
    # Tip dot centers: inset from the extreme tip by TIP_DOT_RADIUS so the
    # dot is visually centered on the rounded cap center.
    tip_center_offset_v = s(ARM_V_REACH - ARM_HALF_WIDTH)   # vertical tips
    tip_center_offset_h = s(ARM_H_REACH - ARM_HALF_WIDTH)   # horizontal tips

    # Top tip
    draw.ellipse([cx - tr, cy - tip_center_offset_v - tr,
                  cx + tr, cy - tip_center_offset_v + tr], fill=white)
    # Bottom tip
    draw.ellipse([cx - tr, cy + tip_center_offset_v - tr,
                  cx + tr, cy + tip_center_offset_v + tr], fill=white)
    # Left tip
    draw.ellipse([cx - tip_center_offset_h - tr, cy - tr,
                  cx - tip_center_offset_h + tr, cy + tr], fill=white)
    # Right tip
    draw.ellipse([cx + tip_center_offset_h - tr, cy - tr,
                  cx + tip_center_offset_h + tr, cy + tr], fill=white)

    return img


# ---------------------------------------------------------------------------
# Build soft inner glow as a blurred radial white halo (RGBA)
# ---------------------------------------------------------------------------
def make_glow(size):
    scale = size / CANVAS
    glow_r = int(GLOW_RADIUS * scale)
    blur_sigma = max(1, int(GLOW_BLUR * scale))

    img = Image.new("RGBA", (size, size), (0, 0, 0, 0))
    cx = cy = size // 2

    # Draw a soft white radial circle that we'll blur
    glow_img = Image.new("RGBA", (size, size), (0, 0, 0, 0))
    gd = ImageDraw.Draw(glow_img)

    # Build concentric circles from outside in for a radial falloff
    steps = 30
    for i in range(steps, 0, -1):
        t = i / steps                          # 1 at edge, ~0 at center
        r = int(glow_r * t)
        alpha = int(255 * GLOW_OPACITY * (1 - t))  # brightest at center
        gd.ellipse(
            [cx - r, cy - r, cx + r, cy + r],
            fill=(255, 255, 255, alpha),
        )

    # Blur for the soft-glow effect
    glow_blurred = glow_img.filter(ImageFilter.GaussianBlur(radius=blur_sigma))
    img.paste(glow_blurred, (0, 0), glow_blurred)
    return img


# ---------------------------------------------------------------------------
# Compose the full icon (background + glow + mark)
# ---------------------------------------------------------------------------
def make_full_icon(size):
    bg   = make_background(size)
    glow = make_glow(size)
    mark = make_mark(size)

    # Composite: background → glow → mark
    result = bg.copy()
    result = Image.alpha_composite(result, glow)
    result = Image.alpha_composite(result, mark)
    return result.convert("RGBA")


# ---------------------------------------------------------------------------
# Save helper
# ---------------------------------------------------------------------------
def save_png(img, path):
    path = Path(path)
    path.parent.mkdir(parents=True, exist_ok=True)
    img.save(str(path), "PNG", optimize=True)
    print(f"  Saved: {path.relative_to(PROJECT_ROOT)}  ({img.size[0]}x{img.size[1]})")


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------
def main():
    print("Loadout Icon Generator")
    print("=" * 50)

    # ── Launcher icons (square, all densities) ─────────────────────────────
    print("\nGenerating launcher icons...")
    for size, folder in LAUNCHER_TARGETS:
        icon = make_full_icon(size)
        save_png(icon, folder / "ic_launcher.png")
        save_png(icon, folder / "ic_launcher_round.png")

    # ── Adaptive icon foreground (1024px, mark on transparent bg) ──────────
    print("\nGenerating adaptive icon foreground layer (1024x1024, transparent bg)...")
    size, folder = FOREGROUND_TARGET
    foreground = make_mark(size)
    save_png(foreground, folder / "ic_launcher_foreground.png")

    print("\nDone. All icons generated successfully.")


if __name__ == "__main__":
    main()

